File: solutions/scala/house/1/src/main/scala/House.scala
import scala.annotation.tailrec

object House {
  private val phraseByNumber: Map[Int, String] = Map(
    1 -> "house that Jack built.",
    2 -> "malt that lay in",
    3 -> "rat that ate",
    4 -> "cat that killed",
    5 -> "dog that worried",
    6 -> "cow with the crumpled horn that tossed",
    7 -> "maiden all forlorn that milked",
    8 -> "man all tattered and torn that kissed",
    9 -> "priest all shaven and shorn that married",
    10 -> "rooster that crowed in the morn that woke",
    11 -> "farmer sowing his corn that kept",
    12 -> "horse and the hound and the horn that belonged to"
    )

  @tailrec
  def recite(nextVerse: Int, endVerse: Int, verses: List[String] = Nil): String =
    if (nextVerse > endVerse) {
      verses.mkString + "\n"
    }
    else {
      recite(nextVerse + 1, endVerse, verses :+ createVerse(nextVerse))
    }

  @tailrec
  private def createVerse(phraseNumber: Int, phrases: List[String] = Nil): String =
    if (phraseNumber == 0) {
      "This is " + phrases.mkString(" ") + "\n"
    }
    else {
      phraseByNumber.get(phraseNumber) match {
        case Some(phrase) => createVerse(phraseNumber - 1, phrases :+ s"the $phrase")
      }
    }
}

File: solutions/python/house/1/house.py
ACTION_BY_SUBJECT = {
    "house that Jack built": "",
    "malt": "lay in",
    "rat": "ate",
    "cat": "killed",
    "dog": "worried",
    "cow with the crumpled horn": "tossed",
    "maiden all forlorn": "milked",
    "man all tattered and torn": "kissed",
    "priest all shaven and shorn": "married",
    "rooster that crowed in the morn": "woke",
    "farmer sowing his corn": "kept",
    "horse and the hound and the horn": "belonged to",
}

def build_verse(subject):
    action = ACTION_BY_SUBJECT[subject]
    if not action:
        return f"the {subject}."

    subjects = list(ACTION_BY_SUBJECT)
    previous_subject = subjects[subjects.index(subject) - 1]
    return (
        f"the {subject} that {action} "
        f"{build_verse(previous_subject)}"
    )

def recite(start_verse, end_verse):
    subjects = list(ACTION_BY_SUBJECT)
    verses = []
    for verse_number in range(start_verse, end_verse + 1):
        subject = subjects[verse_number - 1]
        verses.append(f"This is {build_verse(subject)}")

    return verses